File: python/MergeInBetweenLinkedLists1669.py
# You are given two linked lists: list1 and list2 of sizes n and m respectively.
# Remove list1's nodes from the ath node to the bth node, and put list2 in their place.
# The blue edges and nodes in the following figure indicate the result:
# Build the result list and return its head.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def mergeInBetween(list1: ListNode, a: int, b: int, list2: ListNode) -> ListNode:
    current, i = list1, 0
    while i < a - 1:
        current = current.next
        i += 1
    head = current
    while i <= b:
        current = current.next
        i += 1
    head.next = list2
    while list2.next:
        list2 = list2.next
    list2.next = current
    return list1


if __name__ == "__main__":
    list_1 = ListNode(0, ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))))
    list_2 = ListNode(1000000, ListNode(1000001, ListNode(1000002)))
    list_1 = mergeInBetween(list_1, 3, 4, list_2)
    while list_1:
        print(list_1.val)
        list_1 = list_1.next
    print()
    list_1 = ListNode(0, ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))))
    list_2 = ListNode(1000000, ListNode(1000001, ListNode(1000002)))
    list_1 = mergeInBetween(list_1, 0, 0, list_2)
    while list_1:
        print(list_1.val)
        list_1 = list_1.next
    print()
    list_1 = ListNode(0, ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))))
    list_2 = ListNode(1000000, ListNode(1000001, ListNode(1000002)))
    list_1 = mergeInBetween(list_1, 0, 1, list_2)
    while list_1:
        print(list_1.val)
        list_1 = list_1.next
    print()
    list_1 = ListNode(0, ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))))
    list_2 = ListNode(1000000, ListNode(1000001, ListNode(1000002)))
    list_1 = mergeInBetween(list_1, 1, 4, list_2)
    while list_1:
        print(list_1.val)
        list_1 = list_1.next
    print()
    list_1 = ListNode(0, ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))))
    list_2 = ListNode(1000000, ListNode(1000001, ListNode(1000002)))
    list_1 = mergeInBetween(list_1, 0, 4, list_2)
    while list_1:
        print(list_1.val)
        list_1 = list_1.next

File: java/MergeInBetweenLinkedLists1669.java
/*
You are given two linked lists: list1 and list2 of sizes n and m respectively.
Remove list1's nodes from the ath node to the bth node, and put list2 in their place.
The blue edges and nodes in the following figure indicate the result:
Build the result list and return its head.
*/
public class MergeInBetweenLinkedLists1669 {
    public static void main(String[] args) {
        ListNode list1 = new ListNode(0, new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5))))));
        ListNode list2 = new ListNode(1000000, new ListNode(1000001, new ListNode(1000002)));
        ListNode result = mergeInBetween(list1, 3, 4, list2);
        while (result != null) {
            System.out.println(result.val);
            result = result.next;
        }
    }

    public static ListNode mergeInBetween(ListNode list1, int a, int b, ListNode list2) {
        ListNode current = list1;
        int i = 0;
        for (; i < a - 1; i++) current = current.next;
        ListNode head = current;
        for (; i <= b; i++) current = current.next;
        head.next = list2;
        while (head.next != null) head = head.next;
        head.next = current;
        return list1;
    }

    public static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}
